pkg/webui/components/search-panel/index.js

```diff
@@ -0,0 +1,344 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import classNames from 'classnames'
+import React, { useRef, useEffect, useState, useCallback } from 'react'
+import { defineMessages, useIntl } from 'react-intl'
+
+import { APPLICATION, END_DEVICE, GATEWAY, ORGANIZATION } from '@console/constants/entities'
+
+import Icon, {
+  IconSearch,
+  IconApplication,
+  IconArrowUp,
+  IconArrowDown,
+  IconArrowBack,
+  IconGateway,
+  IconOrganization,
+  IconDevice,
+  IconX,
+} from '@ttn-lw/components/icon'
+import ScrollFader from '@ttn-lw/components/scroll-fader'
+
+import Message from '@ttn-lw/lib/components/message'
+
+import PropTypes from '@ttn-lw/lib/prop-types'
+import sharedMessages from '@ttn-lw/lib/shared-messages'
+import useDebounce from '@ttn-lw/lib/hooks/use-debounce'
+
+import Spinner from '../spinner'
+import Overlay from '../overlay'
+
+import PanelItem from './item'
+import LookingLuke from './luke'
+
+import style from './search-panel.styl'
+
+const m = defineMessages({
+  noResultsFound: 'No results found',
+  noResultsSuggestion: 'Try searching for IDs names, attributes, EUIs or descriptions of:',
+  devices: 'End devices of your{lineBreak}bookmarked applications',
+  searchingEntities: 'Searching applications, gateways, organizations, bookmarks',
+  instructions: 'Use {arrowKeys} to choose, {enter} to select',
+  fetchingTopEntities: 'Fetching top entities…',
+  noTopEntities: 'Seems like you haven’t interacted with any entities yet',
+  noTopEntitiesSuggestion:
+    'Once you created or interacted with entities, they will show up here and you can use this panel to quickly search and navigate to them.',
+})
+
+const categoryMap = {
+  [APPLICATION]: {
+    title: sharedMessages.applications,
+  },
+  [GATEWAY]: {
+    title: sharedMessages.gateways,
+  },
+  [ORGANIZATION]: {
+    title: sharedMessages.organizations,
+  },
+  [END_DEVICE]: {
+    title: sharedMessages.devices,
+  },
+  bookmarks: {
+    title: sharedMessages.bookmarks,
+  },
+  recency: {
+    title: sharedMessages.topEntities,
+  },
+}
+
+const iconMap = {
+  [APPLICATION]: IconApplication,
+  [GATEWAY]: IconGateway,
+  [ORGANIZATION]: IconOrganization,
+  [END_DEVICE]: IconDevice,
+}
+
+const SearchPanel = ({
+  onClose,
+  onSelect,
+  topEntities,
+  searchResults,
+  inline,
+  onQueryChange,
+  searchResultsFetching,
+  topEntitiesFetching,
+  searchQuery,
+}) => {
+  const listRef = useRef()
+  const [selectedIndex, setSelectedIndex] = useState(0)
+  const [query, setQuery] = useState('')
+  const debouncedQuery = useDebounce(query, 350, onQueryChange)
+  const { formatMessage } = useIntl()
+  const lastQuery = useRef('')
+  const isTopEntitiesMode = debouncedQuery === ''
+
+  let items
```

Comment:
From what I've read, it's best to use `useMemo` only for really heavy calculations that take more than ~50ms at which point they would introduce a noticeable delay. This is because \`useMemo\` in itself comes with a memory and computational overhead. I've run some `console.time()` between the calculations, and none of them take more than 0.003ms, so in this case I believe it's better to keep it vanilla and live with the additional computations, which also keeps the code cleaner.

See also: https://maxrozen.com/understanding-when-use-usememo